### PR TITLE
Enhanced file formats support

### PR DESCRIPTION
**Pull request by rocketspacer**
Following up #236 
@madarche Are you alright with the interfaces?
Usage:
```
convict.addParser({ format: 'toml', parse: toml.parse });
convict.addParser({ formats: ['yml', 'yaml'], parse: yaml.safeLoad });
convict.addParsers([
    { format: ['json'], parse: JSON.parse },
    { formats: 'json5', parse: json5.parse }
]);
```

### CODE REVIEW

**Comment by jaawerth** [comment]
You're quick- I was considering doing it too but then realized you'd already done it!

One suggestion: since either way you're just passing to `.concat`, which can take a string or an array, I would suggest replacing the separate `.format` and `.formats` with a single property (maybe `.ext` or `.extension`, since that's what we're actually associating with the parser) that accepts either. That simplifies the API without losing any adaptability or stability.

**Comment by rocketspacer** [comment]
@jaawerth Yeah **extensions** seems like a more appropriate name. How about I make it plural and enforce it to always be an array?
@madarche thoughts?

**Comment by jaawerth** [comment]
I like the use of `concat` to work with both, I would just suggest reducing the separate `formats` and `format` options to only one to both simplify the API and eliminate the (unlikely) possibility of a user accidentally passing both (which would cause the "format" option to be ignored) while leaving it flexible.

Maybe something like this?

```javascript
convict.addFormats = function(formats) {
  formats.forEach(convict.addFormat);
};

convict.addFormat = function(format) {
  if (typeof format.parse !== 'function') {
    throw new Error('Parse function format.parse must be a function');
  }

  const extensions = [].concat(format.extension);
  extensions.forEach(function(ext) {
    if (parsers[ext] && parsers[ext] !== format.parse) {
      throw new Error('Extension ' + ext + ' has already been assigned a parser.');
    }
    parsers[ext] = parser.parse;
  });
};
```

**Comment by madarche** [comment]
Hello @rocketspacer, @jaawerth.

I like the API. And this is neat code. Thanks.

I prefer `extensions` over `formats`. I would prefer a single `addParsers` method, that accept both a single value or an array. That's the JavaScript agile way (compared to Java for example): It enables one to write less code and in this case the legibility is not hindered.

That said, the code of Convict is actually inconsistent in this matter (for example `loadFile` vs `addFormat` and `addFormats`). I prefer the `loadFile` approach, but I would not block any PR based the another approach as long as the code is clean, well tested and the changes are documented in the README.

@rocketspacer I'll be happy to merge your work once you have added documentation in the README and that you've squashed your commits! Thanks again for this very worthy contribution!

**Comment by rocketspacer** [comment]
@madarche I agree with you, less API variation, less documentation, fewer mistakes.

If there can only be a single method. I would prefer one with a singular tone. So I unified it all now under a single method `convict.addParser` and each parser now only accept the `extension` attribute instead of `extension|extensions` as before.

Documentation is updated. If you want, you can squash and merge this PR easily. GitHub supports it now.

BTW, the API documents for `addFormat` is supposed to be `convict.addFormat` not `config.addFormat` right?

File: lib/convict.js [hunk 1]
```diff
@@ -726,23 +726,21 @@ convict.addFormats = function(formats) {
...
+  parsers.forEach(function(parser) {
+    if (!parser) throw new Error('Invalid parser');
+    if (!parser.extension) throw new Error('Missing parser extension');
+    if (!parser.parse) throw new Error('Missing parser parse function');
 
-  for (let i = 0; i < formats.length; i++) {
-    parsers[formats[i]] = parser.parse;
-  }
-};
+    if (typeof parser.parse !== 'function') throw new Error('Invalid parser.parse function');
```
Comment (madarche) [inline lib/convict.js]: I think we should not care with type checks in this context. Adding a parser will fail if the type is not right. Adding type checks is worthy in JavaScript when a situation becomes critical or hard to debug. This is not the case here.

File: lib/convict.js [hunk 2]
```diff
@@ -726,23 +726,21 @@ convict.addFormats = function(formats) {
...
 
-/**
- * Adds new custom file parsers
- */
-convict.addParsers = function(parsers) {
-  parsers.forEach(function(parser) {
-    convict.addParser(parser);
+    const extensions = !Array.isArray(parser.extension) ? [parser.extension] : parser.extension;
+    extensions.forEach(function(extension) {
+      if (typeof extension !== 'string') throw new Error('Invalid parser.extension');
```
Comment (madarche) [inline lib/convict.js]: I think we should not care with type checks in this context. Adding a parser will fail if the type is not right. Adding type checks is worthy in JavaScript when a situation becomes critical or hard to debug. This is not the case here.

**Comment by madarche** [comment]
And you are right about `addFormat` that is supposed to be `convict.addFormat` instead of `config.addFormat`. Better fix it in another PR.

File: lib/convict.js [hunk 3]
```diff
@@ -731,8 +731,8 @@ convict.addParser = function(parsers) {
 
   parsers.forEach(function(parser) {
     if (!parser) throw new Error('Invalid parser');
-    if (!parser.extension) throw new Error('Missing parser extension');
-    if (!parser.parse) throw new Error('Missing parser parse function');
+    if (!parser.extension) throw new Error('Missing parser.extension');
+    if (!parser.parse) throw new Error('Missing parser.parse function');
 
     if (typeof parser.parse !== 'function') throw new Error('Invalid parser.parse function');
```
Comment (madarche) [inline lib/convict.js]: I think we should not care with type checks in this context. Adding a parser will fail if the type is not right. Adding type checks is worthy in JavaScript when a situation becomes critical or hard to debug. This is not the case here.

Comment (rocketspacer) [reply lib/convict.js]: The common consensus is that it should fail if the **parse** function is not a function, either immediately here, or on the next call to **loadFile**.

I still prefer that it fails immediately, as it helps detecting errors sooner. Imagine this code, without immediate type checking there's always a chance for errors to slip into production.
```
// Which parsers.getYamlParser() return undefined
convict.addParser({ extension: 'yml', parse: parsers.getYamlParser() });

const conf = convict({ ... });
if (process.env.NODE_ENV == 'production') convict.loadFile('/etc/config/app.yml');
```
Beside, **addParser** should only be call a few times on startup. So type checking shouldn't hurt any performance.

Comment (madarche) [reply lib/convict.js]: OK

**Comment by madarche** [comment]
Thanks @rocketspacer!

**Comment by madarche** [comment]
And let's wait a couple of days before doing a new minor release with this new feature.

**Comment by madarche** [comment]
@rocketspacer it's on its way with #245

**Pull request merged**